indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/TombstoneHelper.java

```diff
@@ -173,66 +188,66 @@ public Set<DataSegment> computeTombstoneSegmentsForReplace(
-          alignedIntervalEnd = overlap.getEnd();
+        if (Intervals.ETERNITY.getStart().equals(overlap.getStart())) {
+          // Generate a tombstone interval covering the negative eternity interval.
+          buckets = validateAndGetBuckets(buckets + 1, maxBuckets);
+          retVal.add(new Interval(overlap.getStart(), replaceGranularity.bucketStart(overlap.getEnd())));
+        } else if ((Intervals.ETERNITY.getEnd()).equals(overlap.getEnd())) {
+          // Generate a tombstone interval covering the positive eternity interval.
+          buckets = validateAndGetBuckets(buckets + 1, maxBuckets);
+          retVal.add(new Interval(replaceGranularity.bucketStart(overlap.getStart()), overlap.getEnd()));
         } else {
-          alignedIntervalEnd = replaceGranularity.bucketEnd(overlap.getEnd());
-        }
-        long alignedIntervalEndMillis = Math.min(alignedIntervalEnd.getMillis(), JodaUtils.MAX_INSTANT);
-        Interval alignedTombstoneInterval = Intervals.utc(alignedIntervalStartMillis, alignedIntervalEndMillis);
+          // Overlap might not be aligned with the granularity if the used interval is not aligned with the granularity
+          // However when fetching from the iterator, the first interval is found using the bucketStart, which
+          // ensures that the interval is "rounded down" to align with the granularity.
+          // Also, the interval would always be contained inside the "intervalToDrop" because the original REPLACE
+          // is aligned by the granularity, and by extension all the elements inside the intervals to drop would
+          // also be aligned by the same granularity (since intervalsToDrop = replaceIntervals - publishIntervals, and
+          // the right-hand side is always aligned)
+          //
+          // For example, if replaceGranularity is DAY, intervalsToReplace is [2023-02-20, 2023-02-24) (always
+          // aligned with the replaceGranularity), intervalsToDrop is [2023-02-22, 2023-02-24) (they must also be aligned with the replaceGranularity)
+          // If the usedIntervals for the datasource is from [2023-02-22T01:00:00Z, 2023-02-23T02:00:00Z), then
+          // the overlap would be [2023-02-22T01:00:00Z, 2023-02-23T02:00:00Z). When iterating over the overlap we will get
+          // the intervals from [2023-02-22, 2023-02-23) and [2023-02-23, 2023-02-24).
+          IntervalsByGranularity intervalsToDropByGranularity = new IntervalsByGranularity(
+              ImmutableList.of(overlap),
+              replaceGranularity
+          );
 
-        retVal.add(alignedTombstoneInterval);
+          // Helps in deduplication if required. Since all the intervals are uniformly granular, there should be no
+          // no overlap post deduplication
+          HashSet<Interval> intervals = Sets.newHashSet(intervalsToDropByGranularity.granularityIntervalsIterator());
```

Comment:
This is problematic I assume. If the overlap is `ETERNITY`,  `Sets.newHashSet(intervalsToDropByGranularity.granularityIntervalsIterator)` would materialize all the buckets, which would OOM, unless there's a limit on the iterator, which I missed.
The iterator should be limited, or we should walk through the iterator and then pick up the items as we traverse.

I have added a comment about another test case, which I feel would OOM due to this line

Reply:
yes, good point. Updated with tests in https://github.com/apache/druid/pull/15243/commits/5d65ce0ad3391faf3f870c9b88819e923800a715